### PR TITLE
Codemods for 4880 (clerk)

### PR DESCRIPTION
**Pull request by Tobbe**
Codemod for #4880 

Users who are using Clerk Auth can run this codemod to get the fix in #4880 in their projects. Nothing will break if they don't run the codemod, they'll just continue using Clerk as they always have.

---

@jtoar I wasn't sure about the naming/versioning of the codemod. Can you have a look at that please? (And the rest too, if you have time for it.)

### CODE REVIEW

File: packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+export const command = 'update-app'
+export const description =
+  '(v2.1.0->v2.2.0) Updates App.{js,tsx} to use new Clerk auth provider'
+
+export const handler = () => {
+  task('Updating App.{js,tsx}', async () => {
+    const rwPaths = getRWPaths()
+
+    await runTransform({
+      transformPath: path.join(__dirname, 'updateForms.js'),
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts:17]: After renaming the folder:
```suggestion
      transformPath: path.join(__dirname, 'updateClerk.js'),
```

File: packages/codemods/src/codemods/v2.x/updateApp/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Update how we make the ClerkProvider context available to the app
+
+- Update imports
+- Remove old ClerkAuthConsumer and ClerkAuthProvider
+- Insert new ClerkAuthProvider implementation
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.x/updateApp/README.md:5]: Does ClerkAuthProvider actually change? It didn't seem to when I ran it.

File: packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import path from 'path'
+
+import task from 'tasuku'
+
+import getRWPaths from '../../../lib/getRWPaths'
+import runTransform from '../../../lib/runTransform'
+
+export const command = 'update-app'
+export const description =
+  '(v2.1.0->v2.2.0) Updates App.{js,tsx} to use new Clerk auth provider'
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts:10]:
```suggestion
  '(v2.1.0->v2.2.0) Updates App.{js,tsx} to use the new Clerk auth provider'
```

File: packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import path from 'path'
+
+import task from 'tasuku'
+
+import getRWPaths from '../../../lib/getRWPaths'
+import runTransform from '../../../lib/runTransform'
+
+export const command = 'update-app'
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts:8]:
```suggestion
export const command = 'update-clerk'
```

**Review by jtoar — CHANGES_REQUESTED** [review]
Thanks @Tobbe. Still reviewing, but so far I know that we're gonna have to rename a couple folders and files:

- let's rename the folder and files named or prefixed with `updateApp` to match the name of the new command—`update-clerk` (camelcased: `updateClerk`)
- let's also change the name of the folder to `v2.2.x` cause these codemods are for upgrading to v2.2.x

**Review by zackdotcomputer — COMMENTED** [review]
Thanks for handling this @tobbe - I'll be following along mostly as a learning exercise to catch up on how y'all handle codemods so I can help out with them myself next time.

File: packages/codemods/src/codemods/v2.x/updateApp/README.md (hunk 2 above)
Comment (Tobbe) [reply packages/codemods/src/codemods/v2.x/updateApp/README.md:5]: How did you test this? It should change. One easy thing to spot should be that the new provider should use `<ClerkLoaded>` whereas the old one doesn't have that component at all.

Comment (jtoar) [reply packages/codemods/src/codemods/v2.x/updateApp/README.md:5]: My mistake, I see it now.

File: packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts (hunk 1 above)
Comment (Tobbe) [reply packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts]: Done

File: packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts (hunk 4 above)
Comment (Tobbe) [reply packages/codemods/src/codemods/v2.x/updateApp/updateApp.yargs.ts]: Done

File: packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+          (declaration.id.name === 'ClerkAuthProvider' ||
+            declaration.id.name === 'ClerkAuthConsumer')
+        )
+      })
+    ) {
+      j(variableDeclaration).remove()
+    }
+  })
+
+  ast.find(j.VariableDeclaration).insertBefore(`
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts:44]: This `ast.find` needs to be a little more specific. We look for something like `const App = ...` specifically. There could be many other declarations besides the ClerkAuthConsumer or Provider.

File: packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      variableDeclaration.value.declarations.find((declaration) => {
+        return (
+          j.VariableDeclarator.check(declaration) &&
+          j.Identifier.check(declaration.id) &&
+          (declaration.id.name === 'ClerkAuthProvider' ||
+            declaration.id.name === 'ClerkAuthConsumer')
+        )
+      })
+    ) {
+      j(variableDeclaration).remove()
```
Comment (jtoar) [inline packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts:40]: This removes all the comments above too; I'll try to preserve them.

**Review by jtoar — CHANGES_REQUESTED** [review]
Almost ready now. I need to make two more changes:
- I'd like to keep the comments above variable declarations
- The insertion isn't specific enough and inserts before every variable declaration

I'll add those in a commit. It's just taking me a bit to remember how to write these things haha.

**Comment by Tobbe** [comment]
@jtoar I'm happy to do those changes 🙂  I need the practice! Unless you're already on it, of course

**Comment by jtoar** [comment]
@Tobbe feel free! Just felt bad for making you do more work. I'll review speedily this time.

**Comment by Tobbe** [comment]
@jtoar Great, I'll get on it

File: packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts (hunk 6 above)
Comment (jtoar) [reply packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts:40]: @Tobbe to keep comments you could do something like this:

```js
      // assuming `comments` is defined earlier

      if (variableDeclaration.value.comments) {
        comments.concat(
          variableDeclaration.value.comments.map((comment) => comment.value)
        )
      }
```

Comment (jtoar) [reply packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts:40]: And then you'd have to insert them back later

File: packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts (hunk 5 above)
Comment (Tobbe) [reply packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts]: This is what I do now:

```
  const appVariableDeclaration = ast.find(j.VariableDeclaration, {
    declarations: [
      {
        id: { name: 'App' },
      },
    ],
  })
```

File: packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts (hunk 6 above)
Comment (Tobbe) [reply packages/codemods/src/codemods/v2.2.x/updateClerk/updateClerk.ts:44]: What a pain it was to insert them back in! I wanted to put them before the new `ClerkAuthProvider`. But for some reason, when I inserted it using `insertBefore('const ClerkAuthProvider = () => ....')` I couldn't find it with `ast.find(j.VariableDeclaration,`. So I had to construct a separate, proper, `VariableDeclaration`. Attach the comments to that. And finally insert that whole thing into the ast.

**Review by jtoar — APPROVED** [review]
Works beautifully! Thanks @Tobbe

**Pull request merged**